text/0000-network-update-loop.md

```diff
@@ -0,0 +1,706 @@
+      - `INetworkUpdateSystem.NetworkUpdate()`
+- **EarlyUpdate**
+  - RunNetworkEarlyUpdate
+    - `UpdateStage = NetworkUpdateStage.EarlyUpdate`
+    - `foreach (INetworkUpdateSystem in m_EarlyUpdate_Array)`
+      - `INetworkUpdateSystem.NetworkUpdate()`
+  - // other systems
+- **FixedUpdate**
+  - RunNetworkFixedUpdate
+    - `UpdateStage = NetworkUpdateStage.FixedUpdate`
+    - `foreach (INetworkUpdateSystem in m_FixedUpdate_Array)`
+      - `INetworkUpdateSystem.NetworkUpdate()`
+  - ScriptRunBehaviourFixedUpdate
+    - `foreach (MonoBehaviour in m_Behaviours)`
+      - `MonoBehaviour.FixedUpdate()`
+  - // other systems
+- **PreUpdate**
+  - RunNetworkPreUpdate
+    - `UpdateStage = NetworkUpdateStage.PreUpdate`
+    - `foreach (INetworkUpdateSystem in m_PreUpdate_Array)`
+      - `INetworkUpdateSystem.NetworkUpdate()`
+  - // other systems
+- **Update**
+  - RunNetworkUpdate
+    - `UpdateStage = NetworkUpdateStage.Update`
+    - `foreach (INetworkUpdateSystem in m_Update_Array)`
+      - `INetworkUpdateSystem.NetworkUpdate()`
+  - ScriptRunBehaviourUpdate
+    - `foreach (MonoBehaviour in m_Behaviours)`
+      - `MonoBehaviour.Update()`
+  - // other systems
+- **PreLateUpdate**
+  - RunNetworkPreLateUpdate
+    - `UpdateStage = NetworkUpdateStage.PreLateUpdate`
+    - `foreach (INetworkUpdateSystem in m_PreLateUpdate_Array)`
+      - `INetworkUpdateSystem.NetworkUpdate()`
+  - // other systems
+  - ScriptRunBehaviourLateUpdate
+    - `foreach (MonoBehaviour in m_Behaviours)`
+      - `MonoBehaviour.LateUpdate()`
+- **PostLateUpdate**
+  - // other systems
+  - RunNetworkPostLateUpdate
+    - `UpdateStage = NetworkUpdateStage.PostLateUpdate`
+    - `foreach (INetworkUpdateSystem in m_PostLateUpdate_Array)`
+      - `INetworkUpdateSystem.NetworkUpdate()`
+
+As seen above, player loop executes `Initialization` stage and that invokes `NetworkUpdateLoop`'s `RunNetworkInitialization` method which iterates over registered `INetworkUpdateSystem`s in `m_Initialization_Array` and calls `INetworkUpdateSystem.NetworkUpdate()` on them.
+
+In all stages, iterate over a static array and call `NetworkUpdate()` method over `INetworkUpdateSystem` interface pattern is repeated.
+
+## INetworkUpdateSystem
+
+`INetworkUpdateSystem` interface is the ultimate contract between user-defined systems and the `NetworkUpdateLoop`.
+
+### Plain Class Usage
+
+```cs
+public class MyPlainScript : IDisposable, INetworkUpdateSystem
+{
+    public void Initialize()
+    {
+        // extension method calls are equivalent to `NetworkUpdateLoop.RegisterNetworkUpdate(this, NetworkUpdateStage.XXX)`
+        this.RegisterNetworkUpdate(NetworkUpdateStage.EarlyUpdate);
+        this.RegisterNetworkUpdate(NetworkUpdateStage.FixedUpdate);
+        this.RegisterNetworkUpdate(NetworkUpdateStage.Update);
+    }
+
+    // will be called on EarlyUpdate, FixedUpdate and Update stages as registered above
+    public void NetworkUpdate()
+    {
```

Comment:
Even though we can safely access the stage via a singleton since there is only ever one Unity PlayerLoop at once, as a user I probably would find it more intuitive if the arg for what stage this is was passed in to the call else I’d have to google for how to read the singleton value.

Also this helps justify having a dedicated `INetworkUpdateSystem` interface :)